### ISSUE TITLE
Interested in a max-destructure-depth rule in core?

### ISSUE THREAD

**Issue by jasonrhodes** [eslint/eslint#10237]
**Please describe what the rule should do:**

Allows a user to set a "maximum depth" for destructuring objects.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// when the max is set to 2, these warn/fail:

function hello({ person: { name: { first }}}) {
  // etc
}

const {
  style: {
    color: {
      hex
    }
  }
} = theme;
```

```js
// Exceeded maximum destructuring depth at 3 levels (at 1:36)
   function hello({ person: { name: { first }}}) {
// -----------------------------------^

// Exceeded maximum destructuring depth at 3 levels (at 8:7)
         hex
// ------^
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Destructuring is a core part of JS and this is a new style issue related to all JS devs who use destructuring, and can cause readability issues that are sometimes nice to avoid with a rule like this. It could also be a plugin, which I'd be fine with, but I wanted to check first! Thanks! :)

**Note: I already have a basic rule written for this that I'd be happy to clean up and PR if the core team wanted this in core. Otherwise I'll be happy to just build it as a plugin.**
